mlflow/pyfunc/__init__.py

```diff
@@ -78,6 +87,125 @@
+Inference API
+*************
+
+The convention for pyfunc models is to have a ``predict`` method or function with the following
+signature::
+
+    predict(data: pandas.DataFrame) -> [numpy.ndarray | pandas.Series | pandas.DataFrame]
+
+This convention is relied on by other MLflow components.
+
+.. _pyfunc-create-custom:
+
+******************************
+Creating custom Pyfunc models
+******************************
+
+MLflow's persistence modules provide convenience functions for creating models with the
+**pyfunc** flavor in a variety of machine learning frameworks (Scikit-learn, Keras, Pytorch, and
+more); however, they do not cover every use case. For example, you may want to create an MLflow
+model with the **pyfunc** flavor using a framework that MLflow does not natively support.
+Alternatively, you may want to build an MLflow model that executes custom logic when evaluating
+queries, such as preprocessing and postprocessing routines. Therefore, ``mlflow.pyfunc``
+provides utilities for creating **pyfunc** models from arbitrary code and model data.
+
+The :meth:`save_model()` and :meth:`log_model()` methods are designed to support multiple workflows
+for creating custom **pyfunc** models that incorporate custom inference logic and data dependencies.
+Data dependencies are **parameters** and **artifacts** that the inference logic may use to
+evaluate queries. These components are defined as follows:
+
+1. **Parameters**: Python objects. For example, a numpy matrix containing precomputed feature
+   values is a **parameter**.
+2. **Artifacts**: Files or directories, such as serialized Python models or CSVs. For example, a
+   serialized Tensorflow graph is an **artifact**; an MLflow model directory is also an
+   **artifact**.
+
+.. _pyfunc-create-custom-workflows:
+
+Workflows
+#########
+
+:meth:`save_model()` and :meth:`log_model()` support the following workflows:
+
+    1. Collecting **artifacts** and **parameters** to create an MLflow model
+
+       Given a set of **artifact** URIs and **parameters**,
+       :meth:`save_model()` and :meth:`log_model()` can automatically download **artifacts** from
+       their URIs and serialize **parameters** to create an MLflow model directory. **artifacts**
+       and **parameters** are specified by the ``artifacts`` and ``parameters`` method arguments,
+       respectively.
+
+       In this case, you must provide a Python class, called a **model class**, with two
+       methods:
+
+           - ``__init__()``: This method loads and interprets **artifacts** from the MLflow
+             model directory.
+
+             *Note:* **parameters** are *automatically* read from the MLflow model directory as
+             Python objects.
+
+           - ``predict()``: This method leverages **artifacts** and **parameters** to evaluate
+             queries. It must adhere to the :ref:`pyfunc-inference-api`.
+
+       The **model_class** must inherit from :class:`~PythonModel`. It is specified via the
```

Comment:
Maybe we could just say "In this case, you must provide a Python class which inherits from :class:`~PythonModel`, with two methods:" and remove this paragraph.

Reply:
In this case, you must provide a Python class that inherits from :class:`~PythonModel`, with two methods:

Reply:
Removed the paragraph and changed the text to read:

```
       In this case, you must provide a Python class which inherits from :class:`~PythonModel`
       and defines two methods: ``__init__()`` and ``predict()`. This class is specified via the
       ``model_class`` parameter and is automatically serialized and deserialized as a Python
       class object.
```